[Docs] Core Cog Guide

bobloy opened the pull request:
### Type

- [ ] Bugfix
- [ ] Enhancement
- [x] New feature

### Description of the changes

Adds a cog guide for the Core cog and its commands.
A step towards the #1734 goal.

Viewable docs: https://red-discordbot--4597.org.readthedocs.build/en/4597/cog_guides/core.html

bobloy committed:
Core guide initial commit

bobloy committed:
Core guide initial commit

bobloy committed:
Mydata docstrings

bobloy committed:
Mydata docstrings

bobloy committed:
Merge branch 'V3/develop' into cog_guide_core

bobloy committed:
Regenerated docs

bobloy committed:
Regenerated docs

bobloy committed:
Lots more docstrings

bobloy committed:
Lots more docstrings

bobloy committed:
Add document link

bobloy committed:
Add document link

bobloy committed:
Merge branch 'V3/develop' into cog_guide_core

bobloy committed:
Up to load docstrings

bobloy committed:
Up to load docstrings

bobloy committed:
Black formatting

bobloy committed:
Black formatting

bobloy committed:
docstring semantics

bobloy committed:
docstring semantics

bobloy committed:
"cog packages" for loads

bobloy committed:
"cog packages" for loads

bobloy committed:
Refix line lengths

bobloy committed:
Refix line lengths

bobloy committed:
Regenerate docs

bobloy committed:
Regenerate docs

bobloy committed:
Remove link

bobloy committed:
Remove link

bobloy committed:
Regenerate docs with new code

bobloy committed:
Regenerate docs with new code

bobloy committed:
Merge branch 'V3/develop' into cog_guide_core

bobloy committed:
Some small docstrings

bobloy committed:
Some small docstrings

bobloy committed:
Regenerate docs with new .. warning code

bobloy committed:
Regenerate docs with new .. warning code

bobloy committed:
More docstrings into the `set` command

bobloy committed:
More docstrings into the `set` command

bobloy committed:
Bullet lists need blank lines.

bobloy committed:
Bullet lists need blank lines.

bobloy committed:
Regenerate docs

bobloy committed:
Regenerate docs

bobloy committed:
Regenerate docs (fix bullet lists)

bobloy committed:
Regenerate docs (fix bullet lists)

bobloy committed:
set activity statuses docstrings

bobloy committed:
set activity statuses docstrings

bobloy committed:
Docstrings up to helpset pagecharlimit

bobloy committed:
Docstrings up to helpset pagecharlimit

bobloy committed:
Regenerate docs

bobloy committed:
Regenerate docs

bobloy committed:
Maxpages and delete delay docstrings

bobloy committed:
spagetthi ctrl+v

bobloy committed:
tagline and contact

bobloy committed:
A little more contact

bobloy committed:
Up to allowlist

bobloy committed:
Up to command defaultdisablecog

bobloy committed:
Regenerate docs

bobloy committed:
Merge remote-tracking branch 'origin/cog_guide_core' into cog_guide_core

# Conflicts:
#	docs/cog_guides/core.rst
#	redbot/core/core_commands.py

bobloy committed:
Redo custominfo to not be terrible

bobloy committed:
Merge branch 'V3/develop' into cog_guide_core

# Conflicts:
#	redbot/core/core_commands.py

bobloy committed:
Up to default enabled cog

bobloy committed:
More docs

bobloy committed:
Up to command disabledmsg

bobloy committed:
Disabled message stuff was missed

bobloy committed:
Merge branch 'V3/develop' into cog_guide_core

bobloy committed:
Finish the rest of the docstrings

bobloy committed:
Regenerate docs

bobloy committed:
Longer description of the cog

bobloy committed:
Regenerate Docs

bobloy committed:
Merge branch 'V3/develop' into cog_guide_core

bobloy commented:
Finally ready for review.

I did not modify the docstrings for the `licenseinfo` command, but I'm interested in feedback on how that could be improved.

Flame442 commented on redbot/core/core_commands.py:
```diff
@@ -786,9 +802,16 @@ async def mydata_owner_user_deletion_level(self, ctx, level: int):
         """
         Sets how user deletions are treated.
 
+        Example:
+            - `[p]mydata ownermanagement setuserdeletionlevel 1`
+
+        **Arguments:**
+
+        - `<level>` The strictness level for user deletion. See Level guide below.
```
The exact formatting of this is weird, in the very least the empty line after `**Arguments:**` should probably be removed, however it might also be a good idea to bold `Example:`.

(I'm just going to comment on this once, but this formatting is used pretty much in every docstring)

bobloy replied:
I'm open for suggestions on this. The main challenge is the negotiation between RST and how discord displays it. I chose bold because it looked better on discord's side. However, RST requires a blank link to start a list if the colon is not at the end of the line I believe. 

I will put together some other options and offer them for discussion.

bobloy replied:
This has been settled in this poll: https://forms.gle/wiY4kY82d4qZh9358

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+
+Show the current help settings.
+
+.. Warning:: These setting may not be accurate if the default formatter is not in use.
```
```suggestion
.. Warning:: These settings may not be accurate if the default formatter is not in use.
```

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+
+Toggles if ``[p]invite`` should be accessible for the average user.
+
+Bot must me made into a ``Public bot`` in the developer dashboard for public invites to work.
```
```suggestion
The bot must be made into a ``Public bot`` in the developer dashboard for public invites to work.
```

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+Delete data Red has about a user for a user.
+
+This will cause the bot to get rid of or disassociate a lot of non-operational data from the specified user.
+Users have access to different command for this unless they can't interact with the bot at all.
```
```suggestion
Users have access to a different command for this unless they can't interact with the bot at all.
```

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+
+Set to -1 to disable this feature.
+
+This is applied only the current server and not globally.
```
```suggestion
This is only applied to the current server and not globally.
```

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+
+**Arguments:**
+
+- ``[public]`` Whether to send traceback to the current context. Leave blank to send to DMs.
```
```suggestion
- ``[public]`` Whether to send the traceback to the current context. Leave blank to send to your DMs.
```

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+.. note:: To use this cog, load it by typing this::
+
+        [p]load core
+
```
Core does not need to be loaded, this part should be omitted from this guide

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+sent in server text channels.
+Setting the delay to 0 disables this feature.
+
+The bot has to have MANAGE_MESSAGES permission for this to work.
```
I don't believe this is true, if the bot is just deleting its own messages

bobloy replied:
This is based on the decorator `@commands.bot_has_permissions(manage_messages=True)` for that command
I definitely see why it shouldn't be necessary to delete your own messages though. If the decorator could be removed, then it'd make sense to remove it from the docs.

Flame442 replied:
It appears to be there to allow the bot to use the mass delete endpoint. So the permission is required.

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+Using this without a setting will toggle.
+
+Examples:
+    - ``[p]helpset showhidden False`` - Disables embeds on this server.
```
I think you forgot to edit this line when you copy pasted :P

bobloy replied:
Looks like I missed a lot of lines in this section actually. Ended up making more changes than you asked for in this section, including adding docs for the new `showaliases` command.

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+Allows the help command to be sent as a paginated menu instead of separate
+messages.
+
+When enabled, ``[p]help`` will only show one page at a time and use reactions to navigate to other pages.
```
```suggestion
When enabled, ``[p]help`` will only show one page at a time and will use reactions to navigate between pages.
```

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+
+
+Examples:
+    - ``[p]helpset usetick False`` - Disables ticking in DMs.
```
```suggestion
    - ``[p]helpset usetick False`` - Disables ticking when help is sent in DMs.
```

bobloy replied:
Rewording this to `Disables ticking when help is sent to DMs.` to clarify it's talking about where the response from the bot ends up. (i.e. it ticks when I use `[p]help` in a server but the results are sent to my DMs)

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+
+**Arguments:**
+
+- ``[use_tick]`` Whether to tick the help command in DMs. Leave blank to toggle.
```
```suggestion
- ``[use_tick]`` Whether to tick the help command when help is sent in DMs. Leave blank to toggle.
```

bobloy replied:
Same as the above, rewording to `Whether to tick the help command when help is sent to DMs`

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+
+**Description**
+
+This allows the bot to respond indicating the existence of a specific help topic even if the user can't use it.
```
I believe the point of this help setting is to let the user know when a command does *not* exist, ie `[p]help pinh` will respond "Command `pinh` does not exist."

bobloy replied:
This was the existing description for the command. I will look into it further and get back to this one.

bobloy replied:
I believe it is accurate as written. 

As an example:
With `verifyexists` = False, my bot will not respond to `[p]help cleanup duplicates` on the Red support server since it does not have manage messages permissions.
With `verifyexists` = True, my bot will respond to `[p]help cleanup duplicates` despite not being able to use it.

bobloy replied:
This is now being handled in #4892

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+Using this without a setting will toggle.
+
+Examples:
+    - ``[p]helpset verifyexists False`` - Enables showing unusable commands.
```
This description is not accurate

Flame442 replied:
This has still not been fixed. This is unrelated to the other issue, as that text is describing a different behavior which IS accurate (although potentially not intended).

bobloy replied:
This is now resolved, along with a better description of the command.

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+
+Shows info about Red.
+
+See ``[p]custominfo`` to customize.
```
```suggestion
See ``[p]set custominfo`` to customize.
```

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+For that, you need to provide a valid permissions level.
+You can generate one here: https://discordapi.com/permissions.html
+
+Please note that you might need two factor authentication for        some permissions.
```
This section was probably formatted strangely in the docstring, as it is formatted wrong here

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+
+.. code-block:: none
+
+    [p]localallowlist add <users_or_roles...>
```
Here you use `<users_or_roles...>`, but below has only `<users...>`

bobloy replied:
Adjusted all to be `<users_or_roles...>`

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+
+.. code-block:: none
+
+    [p]localallowlist remove <users_or_roles...>
```
Same issue as above

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+
+.. code-block:: none
+
+    [p]localblocklist add <users_or_roles...>
```
Same issue as above

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+
+**Description**
+
+Lists users and roles on the blocklist.
```
```suggestion
Lists users and roles on the server blocklist.
```

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+
+.. code-block:: none
+
+    [p]localblocklist remove <users_or_roles...>
```
Same issue as above

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+
+View the End User Data statements of each 3rd-party module.
+
+This will send an attachment with the End User Data statements of all loaded 3rd party cog.
```
```suggestion
This will send an attachment with the End User Data statements of all loaded 3rd party cogs.
```

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+
+**Arguments:**
+
+- ``<cogs...>`` The cog packages to unload.
```
```suggestion
- ``<cogs...>`` The cog packages to reload.
```

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+To add the keys provide the service name and the tokens as a comma separated
+list of key,values as described by the cog requesting this command.
+
+.. Note:: API tokens are sensitive and should only be used in a private channel or in DM with the bot.
```
```suggestion
.. Note:: API tokens are sensitive, so this command should only be used in a private channel or in DM with the bot.
```

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+    - ``[p]set description`` - Resets the description to the default setting.
+    - ``[p]set description MyBot: A Red V3 Bot``
+
+**Arguments:**
+
+- ``[time]`` The seconds to wait before deleting the command message. Use -1 to disable.
```
This shouldn't be here

bobloy replied:
Added the proper argument name and description.

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+    - ``invisible``
+
+Examples:
+    - ``[p]set status online`` - Clears the activity status.
```
```suggestion
    - ``[p]set status online`` - Clears the status.
```

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+Leaving both streamer and stream_title empty will clear it.
+
+Examples:
+    - ``[p]set competing`` - Clears the activity status.
+    - ``[p]set competing London 2012 Olympic Games``
```
Incorrect

bobloy replied:
Added correct examples here.
Also adjusted the arguments to be in `<>` to indicate required, since that's how each individual argument is presented in the usage.

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+    - ``[p]unignore channel #general`` - Unignores commands in the #general channel.
+    - ``[p]unignore channel`` - Unignores commands in the current channel.
+    - ``[p]ignore channel "General Channels"`` - Use quotes for categories with spaces.
+    - ``[p]unignore channel 356236713347252226`` - Also accepts IDs. Use this method to ignore categories.
```
```suggestion
    - ``[p]unignore channel 356236713347252226`` - Also accepts IDs. Use this method to unignore categories.
```

bobloy replied:
Accepted and an additional ignore needed to be unignore on 3321

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+
+**Arguments:**
+
+- ``<channel>`` The channel to ignore. Can be a category channel.
```
```suggestion
- ``<channel>`` The channel to unignore. Can be a category channel.
```

Flame442 commented on docs/cog_guides/core.rst:
```diff
@@ -0,0 +1,3393 @@
+Remove this server from the ignore list.
+
+Example:
+    - ``[p]unignore server`` - Ignores the current server
```
```suggestion
    - ``[p]unignore server`` - Stops ignoring the current server
```

Flame442 reviewed (changes requested):
This is a long one! Thanks for contributing.

bobloy committed:
Fix some of the suggestions

bobloy committed:
Merge branch 'V3/develop' into cog_guide_core

bobloy committed:
Apply additional suggests and some small rewording.

bobloy committed:
Regenerate docs.

bobloy committed:
Apply suggestions and make additional fixes

bobloy committed:
Regenerate docs

bobloy committed:
Clarifies it's only loaded.

bobloy committed:
Regenerate docs.

bobloy committed:
New formatting, more examples, some fixes

bobloy committed:
Regenerate docs

bobloy committed:
Adds consistant use of dashes to arguments

bobloy committed:
More consistant punctuation

bobloy committed:
Regenerate docs

Jackenmen reviewed (commented):
This PR should update the `.github/labeler.yml` file similarly to how it's done for existing cog guides:
https://github.com/Cog-Creators/Red-DiscordBot/blob/V3/develop/.github/labeler.yml#L1-L5

```yaml
"Category: Bot Core":
  # Source
  - redbot/*
  - redbot/core/__init__.py
  - redbot/core/_sharedlibdeprecation.py
  - redbot/core/bot.py
  - redbot/core/checks.py
  - redbot/core/cli.py
  - redbot/core/cog_manager.py
  - redbot/core/core_commands.py
  - redbot/core/data_manager.py
  - redbot/core/errors.py
  - redbot/core/events.py
  - redbot/core/global_checks.py
  - redbot/core/settings_caches.py
  # Docs
  - docs/framework_apikeys.rst
  - docs/framework_bot.rst
  - docs/framework_cogmanager.rst
  - docs/framework_datamanager.rst
  - docs/framework_events.rst
  - docs/cog_guides/cog_manager_ui.rst
  - docs/cog_guides/core.rst
```

bobloy committed:
Merge branch 'V3/develop' into cog_guide_core

bobloy committed:
Add labeler detection

bobloy commented:
Labeler update has been added.

Flame442 reviewed (changes requested):
There is still one unresolved review comment, and a conflicting file which needs to be fixed. Otherwise, this is looking ready to go!

bobloy committed:
Fix example and use clearing language for helpset verifyexists docstring

bobloy committed:
Regenerate docs

bobloy committed:
Merge branch 'V3/develop' into cog_guide_core

# Conflicts:
#	redbot/core/core_commands.py

bobloy committed:
Add leave example and arguments

bobloy committed:
Regenerate docs

bobloy committed:
Regenerate docs

bobloy committed:
Make embedset more readable

(933aefd3ee9371b3b669cbdd2ac05f6b5532e4bb was the sphinx compatible update)

bobloy committed:
Regenerate docs

bobloy committed:
Black formatting

bobloy committed:
Regenerate docs

bobloy commented:
Merging from master called for additional docstring changes. 

- Added examples and arguments for `[p]leave`
- Updated `[p]embedset` docstring to work with Sphinx and be more readable

Flame442 reviewed (dismissed):
Seems to be all set to go now! Thanks again for the contribution!

bobloy commented:
Re-reviewing this, I missed some additional `embedset` commands that were added in a recent update.
Please don't merge until I've corrected.

bobloy committed:
Adds docstring examples and arguments for multiple `embedset` subcommands

bobloy committed:
Regenerate docs

bobloy commented:
Embedset changes have been added, again ready for review.

Docs can be viewed here: https://red-discordbot--4597.org.readthedocs.build/en/4597/cog_guides/core.html

Flame442 reviewed (approved):
Good to go!
